python/ray/autoscaler/v2/scheduler.py

```diff
@@ -0,0 +1,302 @@
+
+from google.protobuf.json_format import MessageToDict
+
+from ray.autoscaler._private.resource_demand_scheduler import UtilizationScore
+from ray.autoscaler._private.util import NodeType
+from ray.core.generated.autoscaler_pb2 import ResourceRequest
+from ray.core.generated.instance_manager_pb2 import (
+    NodeTypeConfig,
+    ResourceScheduleConfig,
+    ScheduleResourceBundlesReply,
+    ScheduleResourceBundlesRequest,
+)
+
+
+class SchedulingNodeStatus(Enum):
+    """
+    The status of a scheduling node (`SchedulingNode`)
+    """
+
+    # The node is to be launched.
+    TO_LAUNCH = "TO_LAUNCH"
+    # The node is pending to be launched (a pending request, or a pending instance)
+    PENDING = "PENDING"
+    # The node is running.
+    RUNNING = "RUNNING"
+
+
+@dataclass
+class SchedulingNode:
+    """
+    A abstraction of a node that can be scheduled on by the resource scheduler.
+
+    A scheduling node is expected to be used as:
+
+        node  = SchedulingNode.from_node_config(node_config)
+        infeasible = node.try_schedule(requests)
+        score = node.compute_score()
+
+        .... do something with the score ....
+
+    NOTE:
+        One could also extend the scheduling behavior by overriding:
+            1. try_schedule()
+            2. compute_score()
+
+    """
+
+    # Class level node id counter.
+    _next_node_id = 0
+
+    # A unique id simply for differentiating nodes. Not the same as a normal ray node
+    # id.
+    node_id: int
+    # Node type name.
+    node_type: NodeType
+    # Requests committed to be placed on this node.
+    sched_requests: List[ResourceRequest] = field(default_factory=list)
+    # The node's current resource capacity.
+    total_resources: Dict[str, float] = field(default_factory=dict)
+    # The node's available resources.
+    available_resources: Dict[str, float] = field(default_factory=dict)
+    # Node's labels, including static or dynamic labels.
+    labels: Dict[str, str] = field(default_factory=dict)
+    # Status
+    status: SchedulingNodeStatus = SchedulingNodeStatus.TO_LAUNCH
+    # Observability descriptive message for why the node was launched in the
+    # first place.
+    # TODO
+    launch_reason: Optional[str] = None
+
+    @classmethod
+    def from_node_config(
+        cls,
+        node_config: NodeTypeConfig,
+        status: SchedulingNodeStatus = SchedulingNodeStatus.TO_LAUNCH,
+    ) -> "SchedulingNode":
+        """
+        Create a scheduling node from a node config.
+        """
+        return cls(
+            node_id=cls.next_node_id(),
+            node_type=node_config.name,
+            total_resources=dict(node_config.resources),
+            available_resources=dict(node_config.resources),
+            labels=dict(node_config.labels),
+            status=status,
+        )
+
+    def try_schedule(self, requests: List[ResourceRequest]) -> List[ResourceRequest]:
+        """
+        Try to schedule the resource requests on this node.
```

Comment:
IIUC, most of scheduler needs to 1. find the best node given requests & nodes 2. and allocate resources to the best node. But the current API looks like we should 1. allocate resources. 2. find the best node. Is it the intentional design? I feel like it should be actually 2 APIs this way? 

```
scheduleable(List[req]) -> List[feasible_req], List[infeasible_req], score
allocate(List[req])
```
?

Reply:
Yeah, what you describe makes sense, but I think they are pretty similar? 

Since you will essentially need to `allocate` during the `schedulable` process, what I do is basically ask each node to schedule the same requests, and pick a winner. 

I don't have a strong opinion here, I agree what you describe might be more intuitive from an API point of view.